ledger/src/blockstore_meta.rs

```diff
@@ -53,40 +53,177 @@ impl Default for ConnectedFlags {
-    /// The slot height of the block this one derives from.
-    /// The parent slot of the head of a detached chain of slots is None.
-    #[serde(with = "serde_compat")]
-    pub parent_slot: Option<Slot>,
-    /// The list of slots, each of which contains a block that derives
-    /// from this one.
-    pub next_slots: Vec<Slot>,
-    /// Connected status flags of this slot
-    pub connected_flags: ConnectedFlags,
-    /// Shreds indices which are marked data complete.  That is, those that have the
-    /// [`ShredFlags::DATA_COMPLETE_SHRED`][`crate::shred::ShredFlags::DATA_COMPLETE_SHRED`] set.
-    pub completed_data_indexes: BTreeSet<u32>,
+/// Now deprecated format for completed data indexes.
+///
+/// Replaced by [`CompletedDataIndexesV2`].
+pub type CompletedDataIndexesV1 = BTreeSet<u32>;
+/// New format for completed data indexes.
+///
+/// `BTreeSet` is inefficient for de/serialization. We
+/// are migrating to a bitvec to improve that.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
+#[serde(transparent)]
+pub struct CompletedDataIndexesV2 {
+    index: BitVec<MAX_DATA_SHREDS_PER_SLOT>,
+}
+
+// API for CompletedDataIndexesV2 that mirrors BTreeSet<u32> to make migration easier.
+// This allows CompletedDataIndexesV2 to be a drop-in replacement for CompletedDataIndexesV1.
+impl CompletedDataIndexesV2 {
+    #[inline]
+    pub fn iter(&self) -> impl DoubleEndedIterator<Item = u32> + '_ {
+        self.index.iter_ones().map(|i| i as u32)
+    }
+
+    #[inline]
+    pub fn into_iter(&self) -> impl DoubleEndedIterator<Item = u32> + '_ {
+        self.iter()
+    }
+
+    #[inline]
+    pub fn insert(&mut self, index: u32) {
+        self.index.insert_unchecked(index as usize);
+    }
+
+    #[inline]
+    pub fn contains(&self, index: &u32) -> bool {
+        self.index.contains(*index as usize)
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.index.is_empty()
+    }
+
+    #[inline]
+    pub fn range<R>(&self, bounds: R) -> impl DoubleEndedIterator<Item = u32> + '_
+    where
+        R: RangeBounds<u32>,
+    {
+        let start = bounds.start_bound().map(|&b| b as usize);
+        let end = bounds.end_bound().map(|&b| b as usize);
+        self.index.range((start, end)).iter_ones().map(|i| i as u32)
+    }
 }
 
+impl FromIterator<u32> for CompletedDataIndexesV2 {
+    fn from_iter<T: IntoIterator<Item = u32>>(iter: T) -> Self {
+        let index = iter.into_iter().map(|i| i as usize).collect();
+        CompletedDataIndexesV2 { index }
+    }
+}
+
+impl From<CompletedDataIndexesV2> for CompletedDataIndexesV1 {
+    fn from(value: CompletedDataIndexesV2) -> Self {
+        value.into_iter().collect()
+    }
+}
+
+impl From<CompletedDataIndexesV1> for CompletedDataIndexesV2 {
+    fn from(value: CompletedDataIndexesV1) -> Self {
+        value.into_iter().collect()
+    }
+}
+
+/// Helper to define the `SlotMeta` version variants.
+macro_rules! slot_meta_version {
```

Comment:
couldn't this be a  base SlotMetaBase<T> and then two type aliases? What am I missing?